far/vmenu.hpp

```diff
@@ -147,16 +146,80 @@ struct MenuItemEx: menu_item
 	wchar_t AutoHotkey{};
 	size_t AutoHotkeyPos{};
 	std::list<std::pair<int, int>> Annotations;
+
+	int SafeGetFirstAnnotation() const noexcept { return Annotations.empty() ? 0 : Annotations.front().first; }
 };
 
 struct menu_layout;
+enum class item_hscroll_policy;
 
 struct SortItemParam
 {
 	bool Reverse;
 	int Offset;
 };
 
+// Keeps track of the horizontal state of all items.
+// The tracking is best effort. See comments below.
+// Everything is relative to menu_layout::TextArea::first (Left edge).
+class vmenu_horizontal_tracker
```

Comment:
Do we really need all these implementation details exposed in the header?
Maybe forward-declare it and have a unique_ptr to it in the menu?
Yes, it's overhead, but one per menu is nothing.

Reply:
Keeping it by `std::unique_ptr` means an extra allocation. If it's fine with you, I can do that.

Reply:
Ouch! Moved the definition of `vmenu_horizontal_tracker` to vmenu.cpp and changed the type of `m_HorizontalTracker` to `std::unique_ptr<vmenu_horizontal_tracker>`. Got a bunch of `"can't delete an incomplete type"` errors in many other files.

Working on it.

Reply:
Done!